### PR TITLE
Properly close UDPSocket before creating a new one

### PR DESCRIPTION
**Pull request by zachmccormick**
Unless you properly call `close()` on the UDPSocket object, unclosed
sockets will very quickly accumulate. If you are sending stats to
StatsD quickly, then you may actually run out of free ports. This
changes the retry behavior to close the port before trying to open a new
one.

### CODE REVIEW

**Comment by kbogtob** [comment]
Hello @zachmccormick! Thanks again for raising the issue and contributing to the project.

Could you allow me to push on Appboy:fix/issue-141 in order to fix the tests on your PR? Here's the diff anyway if you want to commit it yourself:

```ruby
diff --git a/spec/integrations/connection_edge_case_spec.rb b/spec/integrations/connection_edge_case_spec.rb
index de7cd3f..2a15012 100644
--- a/spec/integrations/connection_edge_case_spec.rb
+++ b/spec/integrations/connection_edge_case_spec.rb
@@ -74,6 +74,9 @@ describe 'Connection edge cases test' do
         allow(fake_socket)
           .to receive(:send)
           .and_raise(IOError.new('closed stream'))
+
+        allow(fake_socket)
+          .to receive(:close)
       end

       context 'when retrying is working' do
@@ -85,6 +88,13 @@ describe 'Connection edge cases test' do
           subject.write('foobar')
         end

+        it 'close the initial socket' do
+          expect(fake_socket)
+            .to receive(:close)
+
+          subject.write('foobar')
+        end
+
         it 'retries on the second opened socket' do
           expect(fake_socket_retry)
             .to receive(:send)
@@ -140,6 +150,9 @@ describe 'Connection edge cases test' do
         allow(fake_socket)
           .to receive(:send)
           .and_raise(Errno::ECONNREFUSED.new('closed stream'))
+
+        allow(fake_socket)
+          .to receive(:close)
       end

       context 'when retrying is working' do
@@ -151,6 +164,13 @@ describe 'Connection edge cases test' do
           subject.write('foobar')
         end

+        it 'close the initial socket' do
+          expect(fake_socket)
+            .to receive(:close)
+
+          subject.write('foobar')
+        end
+
         it 'retries on the second opened socket' do
           expect(fake_socket_retry)
             .to receive(:send)
```

**Comment by kbogtob** [comment]
PR Merged through #143 (which just add tests)
